pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml

```diff
@@ -0,0 +1,343 @@
+        <expected-problems>3</expected-problems>
+        <expected-linenumbers>11,12,13</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            i++;
+            i--;
+            ++i;
+            --i;
+            i += 1;
+            i -= 1;
+            i *= 1; // not OK
+            i /= 1; // not OK
+            i = 5; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: reassigned 'foreach' loop variable</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>6</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            doSomethingWith(i);
+            i = 5; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: increment 'foreach' loop variable, but not as first statement</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>6</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            doSomethingWith(i);
+            i += 1; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>no violation: increment 'foreach' loop variable as first statement</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            i++;
+            doSomethingWith(i);
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>no violation: assign-increment 'foreach' loop variable as first statement</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            i += 1;
+            doSomethingWith(i);
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: increment 'foreach' loop variable multiple times</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>7</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            i++;
+            doSomethingWith(i);
```

Comment:
Would `doSomethingWith(i++)` be flagged here? In fact more generally, would we detect assignments that are in an arbitrary expression, and not just as a simple statement? It seems to me this should be handled as well

Reply:
With the new limited flow analysis `doSomethingWith(i++)` would be reported, unless it happens in a control statement. `doSomethingWith(i = 5)` would be reported, too.

In the current implementation, assignments in expressions aren't handled differently than simple statements.

I'm not sure if we should treat `if (something()) i++;` differently than `if (something()) doSomethingWith(i++)` -- at least in the scope of this rule.
Checking for assignments inside other expressions seems like a good candidate for a separate rule.

Reply:
Hum yes, in fact we already have the rule [AssignmentInOperand](https://pmd.github.io/pmd-6.10.0/pmd_rules_java_errorprone.html#assignmentinoperand). Your rule does well to ignore `if (something()) doSomethingWith(i++)`, it will be flagged by AssignmentInOperand if the user chooses